### PR TITLE
msgpack: friendly binary serialization format

### PR DESCRIPTION
**Pull request by ludamad**
# Description

Right now off the working commit for aztec3. Will be rebased off master when ready.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

**Comment by ludamad** [comment]
@codygunton Can I get a review on this next week? If you prefer to get this in rather than having it sit on an aztec branch it's totally understandable. So far I made sure it has no effect on barretenberg - it is entirely for circuits to use ATM. I have exercised msgpack in a single test.

File: cpp/src/barretenberg/dsl/acir_format/sha256_constraint.hpp [hunk 1]
```diff
@@ -10,13 +10,17 @@ struct Sha256Input {
     uint32_t num_bits;
 
     friend bool operator==(Sha256Input const& lhs, Sha256Input const& rhs) = default;
+    // for serialization, update with any new fields
```
Comment (ludamad) [inline cpp/src/barretenberg/dsl/acir_format/sha256_constraint.hpp:13]: taking feedback on whether this is better as a tight MSGPACK
or
void msgpack(auto archive) {
   archive(NVP(fields...));
}
NVP is needed though to get field names, at a minimum

File: cpp/src/barretenberg/common/msgpack.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+This approach assumes 1. all members are default constructible 2. you give it all members 3. all members are writable references
+
+This method maps the object's properties (e.g., `composer_type`, `circuit_size`, etc.) to their respective keys in the serialized data.
+
+### Array-like Objects
+
+For array-like objects, define the `msgpack` method similarly:
+
+```cpp
```
Comment (codygunton) [inline cpp/src/barretenberg/common/msgpack.hpp]: This is literally the same as in the map-like case, right?

File: cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+// Note: heavy header due to serialization logic, don't include if msgpack.hpp will do
+#include <string>
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+#include <cassert>
+#define MSGPACK_NO_BOOST
+#include <msgpack.hpp>
```
Comment (codygunton) [inline cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp]: Any particular reason to use this form of `include` here?

File: cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#include "concepts.hpp"
+
+namespace msgpack {
+    template <typename Tuple, std::size_t... Is> auto drop_keys_impl(Tuple&& tuple, std::index_sequence<Is...>)
+    {
+        // Expand 0 to n/2 to 1 to n+1 (increments of 2)
```
Comment (codygunton) [inline cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp]: Took me a second to understand this comment, but: $[0,n/2]$ ($n$ even) is transformed to the odd numbers in $[1, n+1]$ :heavy_check_mark:

File: cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        T{args...};
+    };
+}
+
+namespace msgpack::adaptor {
+// reads structs with msgpack() method from a JSON-like dictionary
+template <HasMsgPack T> struct convert<T> {
+    msgpack::object const& operator()(msgpack::object const& o, T& v) const
+    {
+        static_assert(std::is_default_constructible_v<T>, "MSGPACK requires default-constructible types");
```
Comment (codygunton) [inline cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp]: tldr for why we require default constructible here?

File: cpp/src/barretenberg/ecc/fields/field_impl.hpp [hunk 6]
```diff
@@ -607,4 +606,29 @@ template <class T> constexpr field<T> field<T>::multiplicative_generator() noexc
     return target;
 }
 
+// For serialization
+template <class Params> void field<Params>::msgpack_pack(auto& packer) const
+{
+    auto adjusted = from_montgomery_form();
+    uint64_t bin_data[4] = {
+        htonll(adjusted.data[3]), htonll(adjusted.data[2]), htonll(adjusted.data[1]), htonll(adjusted.data[0])
```
Comment (codygunton) [inline cpp/src/barretenberg/ecc/fields/field_impl.hpp:620]: Are the conversions here required by msgpack? Or by some particular application? "Internet network byte order" is new to me :)

File: cpp/src/barretenberg/plonk/proof_system/verification_key/verification_key.hpp [hunk 7]
```diff
@@ -83,6 +95,24 @@ struct verification_key {
     bool contains_recursive_proof = false;
     std::vector<uint32_t> recursive_proof_public_input_indices;
     size_t program_width = 3;
+
+    // for serialization: update up with new fields
```
Comment (codygunton) [inline cpp/src/barretenberg/plonk/proof_system/verification_key/verification_key.hpp]: Is this being tested elsewhere? Ditto for proof serialization.

File: cpp/src/barretenberg/dsl/acir_format/sha256_constraint.hpp (hunk 1 above)
Comment (codygunton) [reply cpp/src/barretenberg/dsl/acir_format/sha256_constraint.hpp:13]: OK so, roughly, `MSGPACK(inputs, result)` defines a function template `msgpack` which will take a packing function and try to pass it `"inputs":inputs, "result":result` as 'key-value pairs', and you're wondering whether it's better to be verbose and avoid the macro? I think the macro is ok, but I might name it something like `MSGPACK_FIELDS` or `TO_SERIALIZE` for an additional bit of clarity.

**Review by codygunton — APPROVED** [review]
Thanks for this, looks good, feels like a big step forward.

File: cpp/src/barretenberg/plonk/proof_system/verification_key/verification_key.hpp (hunk 7 above)
Comment (ludamad) [reply cpp/src/barretenberg/plonk/proof_system/verification_key/verification_key.hpp]: I was thinking a bunch of these should be exercised, but figured I'd get some approval in principle for the spike first

Starting to get some soft confirms, will do, maybe after buy-in from Charlie

File: cpp/src/barretenberg/common/msgpack.hpp (hunk 2 above)
Comment (ludamad) [reply cpp/src/barretenberg/common/msgpack.hpp]: You are correct, the array case was simplified in favour of later serializing everything as arrays if we need the efficiency

File: cpp/src/barretenberg/dsl/acir_format/sha256_constraint.hpp (hunk 1 above)
Comment (ludamad) [reply cpp/src/barretenberg/dsl/acir_format/sha256_constraint.hpp:13]: I do like the word MSGPACK in there as I feel it needs some good correlation with the error stack traces people might face
MSGPACK_FIELDS isn't bad, I'll consider it

File: cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp (hunk 5 above)
Comment (ludamad) [reply cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp]: The msgpack unpacking code simply runs .msgpack() after default constructing. I'll add a rationale right in the error

File: cpp/src/barretenberg/ecc/fields/field_impl.hpp (hunk 6 above)
Comment (ludamad) [reply cpp/src/barretenberg/ecc/fields/field_impl.hpp:620]: This is required by barretenberg :) I made sure the buffer formats were exactly the same. 'network byte order' just means big endian in practicality, it's how bberg already does it in serialize.hpp. But yeah overall just replicating bberg format exactly because the typescript code still supports both buffer formats, so made sense to align them

**Comment by ludamad** [comment]
Thanks a lot for the approval and feedback. I will clean up soon, likely after some demoing

File: cpp/src/barretenberg/ecc/fields/field_impl.hpp (hunk 6 above)
Comment (codygunton) [reply cpp/src/barretenberg/ecc/fields/field_impl.hpp:620]: Alright, thanks for explaining.

File: cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp (hunk 3 above)
Comment (ludamad) [reply cpp/src/barretenberg/common/msgpack/struct_map_impl.hpp]: It's an external library

File: cpp/src/barretenberg/honk/sumcheck/sumcheck.test.cpp [hunk 8]
```diff
@@ -96,7 +96,6 @@ TEST(Sumcheck, PolynomialNormalization)
     std::array<FF, multivariate_n> id_3;
     std::array<FF, multivariate_n> lagrange_first;
     std::array<FF, multivariate_n> lagrange_last;
-    std::array<FF, multivariate_n> pow_zeta;
```
Comment (ludamad) [inline cpp/src/barretenberg/honk/sumcheck/sumcheck.test.cpp:99]: unused

File: cpp/src/barretenberg/ecc/groups/element_impl.hpp [hunk 9]
```diff
@@ -925,7 +925,6 @@ element<Fq, Fr, T> element<Fq, Fr, T>::random_coordinates_on_curve(numeric::rand
     Fq yy;
     Fq x;
     Fq y;
-    Fq t0;
```
Comment (ludamad) [inline cpp/src/barretenberg/ecc/groups/element_impl.hpp:928]: unused

**Comment by ludamad** [comment]
@dbanks12 @charlielye this should be ready for final review. This isn't everything I want to do in bb, but it's everything needed to unblock the work in circuits

File: cpp/src/barretenberg/plonk/proof_system/prover/c_bind.cpp [hunk 10]
```diff
@@ -1,13 +1,17 @@
 #include "prover.hpp"
 #include "barretenberg/env/data_store.hpp"
 #include "barretenberg/env/crs.hpp"
+#include "barretenberg/serialize/test_helper.hpp"
+#include "barretenberg/serialize/raw_pointer.hpp"
 
-#define WASM_EXPORT __attribute__((visibility("default")))
+#define WASM_EXPORT extern "C" __attribute__((visibility("default")))
```
Comment (ludamad) [inline cpp/src/barretenberg/plonk/proof_system/prover/c_bind.cpp:7]: Note it's better to have extern C in our export declare than have extern C blocks IMO

File: cpp/src/barretenberg/plonk/proof_system/prover/c_bind.cpp [hunk 11]
```diff
@@ -50,6 +56,14 @@ WASM_EXPORT void* test_env_load_prover_crs(size_t num_points)
 
 using WasmProver = plonk::UltraProver;
 
+typedef RawPointer<plonk::UltraProver> WasmProverPtr;
+
+// TODO(AD): Currently just a motivating example, TODO bind rest of library
+CBIND(prover_process_queue2, [](WasmProverPtr prover) {
```
Comment (ludamad) [inline cpp/src/barretenberg/plonk/proof_system/prover/c_bind.cpp:62]: @charlielye this is what I'm envisioning raw pointer binds will look like

File: cpp/src/barretenberg/common/log.hpp [hunk 12]
```diff
@@ -18,23 +18,10 @@
...
-{
-    os << first;
-    format_chain(os, args...);
-}
-
 template <typename... Args> std::string format(Args... args)
 {
     std::ostringstream os;
-    format_chain(os, args...);
+    ((os << args), ...);
```
Comment (dbanks12) [inline cpp/src/barretenberg/common/log.hpp:24]: What does this do?

File: cpp/src/barretenberg/ecc/curves/bn254/g1.hpp [hunk 13]
```diff
@@ -17,4 +17,11 @@ struct Bn254G1Params {
...
-} // namespace barretenberg
\ No newline at end of file
+
+} // namespace barretenberg
+
+// specialize the name in msgpack schema generation
+inline std::string msgpack_schema_name(barretenberg::g1::affine_element const&)
+{
+    return "G1AffineElement";
+}
```
Comment (dbanks12) [inline cpp/src/barretenberg/ecc/curves/bn254/g1.hpp]: Why is it that we want to specialize these names? Is it related to naming in aztec circuits or TS?

File: cpp/src/barretenberg/ecc/fields/field_impl.hpp (hunk 6 above)
Comment (dbanks12) [inline cpp/src/barretenberg/ecc/fields/field_impl.hpp]: Comments would be helpful here, otherwise this just looks like voodoo.

File: cpp/src/barretenberg/serialize/cbind.hpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+template <typename T> constexpr auto param_tuple()
+{
+    return typename decltype(get_func_traits<T>())::Args{};
+}
+
+inline void msgpack_cbind_impl(
+    auto func, const uint8_t* input_in, size_t input_len_in, uint8_t** output_out, size_t* output_len_out)
+{
+    auto params = param_tuple<decltype(func)>();
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/cbind.hpp]: I'd add comments for param_tuple's definition and usage so it is clear what it is doing. Something like "Extracts a function's parameter types and creates a tuple with entries matching those types. Useful for unpacking a buffer into a tuple of parameters that can then be applied to the function."

File: cpp/src/barretenberg/serialize/cbind.hpp [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+// 2. cname##__schema function: This creates a JSON schema of the function's input arguments and return type.
+#define CBIND(cname, func)                                                                                             \
+    WASM_EXPORT void cname(const uint8_t* input_in, size_t input_len_in, uint8_t** output_out, size_t* output_len_out) \
+    {                                                                                                                  \
+        msgpack_cbind_impl(func, input_in, input_len_in, output_out, output_len_out);                                  \
+    }                                                                                                                  \
+    WASM_EXPORT void cname##__schema(uint8_t** output_out, size_t* output_len_out)                                     \
+    {                                                                                                                  \
+        msgpack_cbind_schema_impl(func, output_out, output_len_out);                                                   \
+    }
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/cbind.hpp:102]: beautiful :+1:

File: cpp/src/barretenberg/serialize/msgpack.hpp [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+    // Convert the buffer data to a string and return it
+    return { output, buffer.size() };
+```
+
+e.g. unpacking
+
+```
+    msgpack::unpack((const char*)encoded_data, encoded_data_size).get().convert(*value);
+```
+*/
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack.hpp]: More comments please .... :rofl: kidding, nice :fire:

File: cpp/src/barretenberg/serialize/msgpack.hpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
+#pragma once
+/* Minimal header for msgpack. This should be included, *not* msgpack.hpp directly.
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack.hpp]: What does this mean? I'm probably misunderstanding, but I read this as "msgpack.hpp should be included, not msgpack.hpp directly".

File: cpp/src/barretenberg/serialize/msgpack.hpp [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+#include <type_traits>
+#include "msgpack_impl/name_value_pair_macro.hpp"
+#include "msgpack_impl/concepts.hpp"
+
+// Helper for above documented syntax
+#define MSGPACK_FIELDS(...)                                                                                            \
+    void msgpack(auto pack_fn)                                                                                         \
+    {                                                                                                                  \
+        pack_fn(NVP(__VA_ARGS__));                                                                                     \
+    }
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack.hpp:105]: Consider adding a comment here as this is voodoo to me. Maybe unimportant if this is just a silly helper.

File: cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+     * @param schema_name The CPP type.
+     * @param msgpack_name The msgpack type.
+     */
+    void pack_alias(const std::string& schema_name, const std::string& msgpack_name)
+    {
+        pack_array(2);
+        pack("alias");
+        pack_array(2);
+        pack(schema_name);
+        pack(msgpack_name);
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp]: what is 2 here?

File: cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    msgpack::check_msgpack_usage(object);
+    // Encode as map
+    const_cast<T&>(object).msgpack([&](auto&... args) {
+        size_t kv_size = sizeof...(args);
+        packer.pack_map(uint32_t(1 + kv_size / 2));
+        packer.pack("__typename");
+        packer.pack(type);
+        _schema_pack_map_content(packer, args...);
+    });
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp]: More explanation here would be helpful

File: cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp (hunk 19 above)
Comment (dbanks12) [reply cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp]: same question in a few other spots

**Review by dbanks12 — APPROVED** [review]
Great work! Looks good to me. A lot of the code was familiar from a previous review I did (maybe on a PoC PR in the circuits repo?).

I left a few comments, but they are mostly asking for more clarification comments. Since this is a large PR and we want to get this in ASAP, this can be merged as-is IMO, and then my minor comments can be addressed later.

File: cpp/src/barretenberg/common/log.hpp (hunk 12 above)
Comment (ludamad) [reply cpp/src/barretenberg/common/log.hpp:24]: What the previous code did. Run os << args on each args. It's a funky syntax, but it lets us simplify the helpers. This is the way to 'loop' Args... args parameters

File: cpp/src/barretenberg/ecc/curves/bn254/g1.hpp (hunk 13 above)
Comment (ludamad) [reply cpp/src/barretenberg/ecc/curves/bn254/g1.hpp]: It is for Typescript. As for why, it is largely related to C++ templates. Recall that this would expand to some affine_element<MetaStruct<Parameters<>>>. It's quite hard to parse that on the TS side, and then if any refactor were done, we would have to update brittlely based on the type name. It felt more stable to put the maintenance burden on the coder. I didn't know how much to capture of that in the comment

File: cpp/src/barretenberg/serialize/msgpack.hpp (hunk 17 above)
Comment (ludamad) [reply cpp/src/barretenberg/serialize/msgpack.hpp]: Ill fix this, this should be included not just msgpack.hpp which resolves to the msgpack library itself
